README.md
# Modeling match results in the Dutch Eredivisie using a hierarchical Bayesian Poisson model

> *Copyright 2019 [Piet Stam](http://www.pietstam.nl). The README.md documentation and the .Rmd, .R and .html files are licensed under the Creative Commons [Attribution 4.0 International license](http://creativecommons.org/licenses/by/4.0/).* 

## Quick summary
`r-bayesian-football-odds` is based on the works of [Rasmus Baath](http://www.sumsar.net/blog/2013/07/modeling-match-results-in-la-liga-part-one/). Rasmus Baath submitted his code to the [UseR 2013 data analysis contest](https://www.r-project.org/conferences/useR-2013/) and licensed it under the Creative Commons [Attribution 3.0 Unported license](http://creativecommons.org/licenses/by/3.0/). 

## The original works
Rasmus predicted the results of the 50 last matches of the 2012/2013 Spanish LaLiga season. He used data of the 2008/09-2012/13 seasons (5 seasons in total) to estimate his regression model in a [Bayesian](https://en.wikipedia.org/wiki/Bayes_estimator) way. See [this thread](https://stats.stackexchange.com/questions/252577/bayes-regression-how-is-it-done-in-comparison-to-standard-regression) for an intuitive explanation of the difference between the bayesian approach and the conventional approaches of linear regression and maximum likelihood.

## My application
I slightly adpated his code to predict the results of the last two competition rounds (that is, the last 18 matches) of the 2018/2019 Dutch Eredivisie season. These predictions are based on soccer match data of the 2014/15-2018/19 seasons (5 seasons in total). The source of these data is [here](http://www.football-data.co.uk/netherlandsm.php). Out of the three model specifications that Rasmus developed, I used the most sophisticated model that allowed for year-to-year variability in team skill (called "iteration 3" by Rasmus).

## The results
If you want to dive into the results of my application to the Dutch Eredivisie right away, you may click on the [NL_eredivisie_2014_2019.md](https://github.com/pjastam/r-bayesian-football-odds/blob/master/NL_eredivisie_2014_2019.md) file in this GitHub repository. The last section in this file contains the results of my one-off betting experience with VERSION 1.0.


NEWS.md
# CHANGES IN r-bayesian-football-odds VERSION 1.3

## NEW FEATURES

- We now make use of the `runjags` package to perform parallel computations, i.e. using multiple cores of our machine. 

## BUG FIXES

- We applied the TOC as output option in the YAML header of the R Markdown (.Rmd) file instead of the old fashioned way of using HTML code to diplay a list of contents.

- Easy fix to have the Probability Graph Model (PGM) displayed at GitHub. Until this fix it was not the case.

- We changed the output format of the R Markdown (.Rmd) file from html_document to github_document.

- As a consequence of changing the output format to github_document, we also had to change the code of the math statements. Using a helper function written by Ogan Mancarcı (https://github.com/STAT545-UBC/Discussion/issues/102#issuecomment-142482040), we used inline R code to this end.

# CHANGES IN r-bayesian-football-odds VERSION 1.2

## NEW FEATURES

- I added a last section to the Rmd file, showing the results of my one-off betting experience with VERSION 1.0.

## BUG FIXES

- The path to the data directory in the `raw_data` chunk was wrongly called "test". We replaced it by "data".

- We found that one record was missing in the N120182019.csv file, because the total amount of records in this file was one less than 306, which is the total number of matches in one football season. We added the missing record of the Zwolle-Groningen match at April 29th, 2019.

- If the contents of the data direcory are changed, it appears that the cache is not updated because `cache=1` or `cache=2` don't work with the `cache.extra` option, see https://github.com/yihui/knitr/issues/994 and https://yihui.name/knitr/demo/cache/#more-granular-cache. This problem is solved by splitting the code chunck into two code chunks, one of which does nothing more that scanning the status of the data directory.

# CHANGES IN r-bayesian-football-odds VERSION 1.1

## NEW FEATURES

- Parameterized the start year and end year when reading the raw data, plotting the histograms and estimating the team skill parameters.

- Added caching of the MCMC results. The cache updates automatically when the raw data, the processed data and/or the JAGS model descrition change. Otherwise, knitting uses the cached MCMC results.

- Added the hook hook_purl() to write the code chunks to an R script when knitting. See ?hook_purl and the example at https://stackoverflow.com/questions/27462018/knitr-do-not-purl-a-chunk#27488012.

## MINOR CHANGES

- Added chunck names.

- The script tag in the mathematical formulas is replaced by the $$ tag, in line with the original code by Rasmus.

- Changed order of sections in html output.

# CHANGES IN r-bayesian-football-odds VERSION 1.0

## NEW FEATURES

- first version of r-bayesian-football-odds: this is the version that I announced on my blog at https://pietstam.nl/blog/2019/05/10/bayesian-football-odds and posted about my one-off betting experience on LinkedIn (in Dutch): https://www.linkedin.com/pulse/de-bal-rond-piet-stam/  (short summary in English: 3 euro stake earned 4.15 euro, so a profit margin of 1.15 / 3 = 38%)


functions/plotPost.R
plotPost = function( paramSampleVec , credMass=0.95 , compVal=NULL ,
           HDItextPlace=0.7 , ROPE=NULL , yaxt=NULL , ylab=NULL ,
           xlab=NULL , cex.lab=NULL , cex=NULL , xlim=NULL , main=NULL ,
           col=NULL , border=NULL , showMode=F , showCurve=F , breaks=NULL , 
                     ... ) {
    # Override defaults of hist function, if not specified by user:
    # (additional arguments "..." are passed to the hist function)
    if ( is.null(xlab) ) xlab="Parameter"
    if ( is.null(cex.lab) ) cex.lab=1.5
    if ( is.null(cex) ) cex=1.4
    if ( is.null(xlim) ) xlim=range( c( compVal , paramSampleVec ) )
    if ( is.null(main) ) main=""
    if ( is.null(yaxt) ) yaxt="n"
    if ( is.null(ylab) ) ylab=""
    if ( is.null(col) ) col="skyblue"
    if ( is.null(border) ) border="white"

    postSummary = matrix( NA , nrow=1 , ncol=11 , 
                          dimnames=list( c( xlab ) , 
                                         c("mean","median","mode",
                                           "hdiMass","hdiLow","hdiHigh",
                                           "compVal","pcGTcompVal",
                                           "ROPElow","ROPEhigh","pcInROPE")))              
    postSummary[,"mean"] = mean(paramSampleVec)
    postSummary[,"median"] = median(paramSampleVec)
    mcmcDensity = density(paramSampleVec)
    postSummary[,"mode"] = mcmcDensity$x[which.max(mcmcDensity$y)]

    source("functions/HDIofMCMC.R")
    HDI = HDIofMCMC( paramSampleVec , credMass )
    postSummary[,"hdiMass"]=credMass
    postSummary[,"hdiLow"]=HDI[1]
    postSummary[,"hdiHigh"]=HDI[2]

    # Plot histogram.
    if ( is.null(breaks) ) {
    breaks = c( seq( from=min(paramSampleVec) , to=max(paramSampleVec) ,
                   by=(HDI[2]-HDI[1])/18 ) , max(paramSampleVec) )
    }
    if ( !showCurve ) {
      par(xpd=NA)
      histinfo = hist( paramSampleVec , xlab=xlab , yaxt=yaxt , ylab=ylab ,
                       freq=F , border=border , col=col ,
                       xlim=xlim , main=main , cex=cex , cex.lab=cex.lab ,
                       breaks=breaks , ... )
    }
    if ( showCurve ) {
      par(xpd=NA)
      histinfo = hist( paramSampleVec , plot=F )
      densCurve = density( paramSampleVec , adjust=2 )
      plot( densCurve$x , densCurve$y , type="l" , lwd=5 , col=col , bty="n" ,
            xlim=xlim , xlab=xlab , yaxt=yaxt , ylab=ylab ,
            main=main , cex=cex , cex.lab=cex.lab , ... )
    }
    cenTendHt = 0.9*max(histinfo$density)
    cvHt = 0.7*max(histinfo$density)
    ROPEtextHt = 0.55*max(histinfo$density)
    # Display mean or mode:
    if ( showMode==F ) {
        meanParam = mean( paramSampleVec )
        text( meanParam , cenTendHt ,
              bquote(mean==.(signif(meanParam,3))) , adj=c(.5,0) , cex=cex )
    } else {
        dres = density( paramSampleVec )
        modeParam = dres$x[which.max(dres$y)]
        text( modeParam , cenTendHt ,
              bquote(mode==.(signif(modeParam,3))) , adj=c(.5,0) , cex=cex )
    }
    # Display the comparison value.
    if ( !is.null( compVal ) ) {
      cvCol = "darkgreen"
      pcgtCompVal = round( 100 * sum( paramSampleVec > compVal )
                            / length( paramSampleVec )  , 1 )
       pcltCompVal = 100 - pcgtCompVal
       lines( c(compVal,compVal) , c(0.96*cvHt,0) ,
              lty="dotted" , lwd=1 , col=cvCol )
       text( compVal , cvHt ,
             bquote( .(pcltCompVal)*"% < " *
                     .(signif(compVal,3)) * " < "*.(pcgtCompVal)*"%" ) ,
             adj=c(pcltCompVal/100,0) , cex=0.8*cex , col=cvCol )
      postSummary[,"compVal"] = compVal
      postSummary[,"pcGTcompVal"] = ( sum( paramSampleVec > compVal ) 
                                  / length( paramSampleVec ) )
    }
    # Display the ROPE.
    if ( !is.null( ROPE ) ) {
      ropeCol = "darkred"
       pcInROPE = ( sum( paramSampleVec > ROPE[1] & paramSampleVec < ROPE[2] )
                            / length( paramSampleVec ) )
       lines( c(ROPE[1],ROPE[1]) , c(0.96*ROPEtextHt,0) , lty="dotted" , lwd=2 ,
              col=ropeCol )
       lines( c(ROPE[2],ROPE[2]) , c(0.96*ROPEtextHt,0) , lty="dotted" , lwd=2 ,
              col=ropeCol)
       text( mean(ROPE) , ROPEtextHt ,
             bquote( .(round(100*pcInROPE))*"% in ROPE" ) ,
             adj=c(.5,0) , cex=1 , col=ropeCol )
      
      postSummary[,"ROPElow"]=ROPE[1] 
      postSummary[,"ROPEhigh"]=ROPE[2] 
      postSummary[,"pcInROPE"]=pcInROPE
    }
    # Display the HDI.
    lines( HDI , c(0,0) , lwd=4 )
    text( mean(HDI) , 0 , bquote(.(100*credMass) * "% HDI" ) ,
          adj=c(.5,-1.7) , cex=cex )
    text( HDI[1] , 0 , bquote(.(signif(HDI[1],3))) ,
          adj=c(HDItextPlace,-0.5) , cex=cex )
    text( HDI[2] , 0 , bquote(.(signif(HDI[2],3))) ,
          adj=c(1.0-HDItextPlace,-0.5) , cex=cex )
    par(xpd=F)
    #
    return( postSummary )
}


functions/Read_N20142019.R
rm(list = ls())
library(readr)

eredivisie <- NULL
for (i in 2014:2018) {
  nam <- paste("N1",i,i+1,sep="")
  N2 <- assign(nam, read_csv(paste("data/N1",i,i+1,".csv", sep=""), 
                       col_types = cols_only(Date = col_datetime(format = "%d/%m/%y"), 
                                             FTHG = col_integer(), 
                                             FTAG = col_integer(), 
                                             HomeTeam = col_character(),
                                             AwayTeam = col_character())))
  N2 <- N2[complete.cases(N2[,c("Date","HomeTeam","AwayTeam")])==TRUE,]
  Season <- paste(i,"/",i+1, sep="")
  Week <- as.integer(floor(difftime(N2$Date,N2$Date[1],units="weeks"))+1)
  N2 <- cbind(Season,Week,N2)
  eredivisie <- rbind(eredivisie,N2)
}
names(eredivisie) <- c('Season','Week','Date','HomeTeam','AwayTeam','HomeGoals','AwayGoals')
save(eredivisie, file="results/eredivisie.RData")

functions/Import_Data_Eredivisie.R
library(readr)

eredivisie <- NULL
for (i in from_year:(to_year-1)) {
  nam <- paste("N1",i,i+1,sep="")
  N2 <- assign(nam, read_csv(paste("data/N1",i,i+1,".csv", sep=""), 
                       col_types = cols_only(Date = col_datetime(format = "%d/%m/%y"), 
                                             FTHG = col_integer(), 
                                             FTAG = col_integer(), 
                                             HomeTeam = col_character(),
                                             AwayTeam = col_character())))
  N2 <- N2[complete.cases(N2[,c("Date","HomeTeam","AwayTeam")])==TRUE,]
  Season <- paste(i,"/",i+1, sep="")
  Week <- as.integer(floor(difftime(N2$Date,N2$Date[1],units="weeks"))+1)
  N2 <- cbind(Season,Week,N2)
  eredivisie <- rbind(eredivisie,N2)
  rm(N2)
}
names(eredivisie) <- c('Season','Week','Date','HomeTeam','AwayTeam','HomeGoals','AwayGoals')

NL_eredivisie_2014_2019.R
## ----setup, include=FALSE-----------------------------------------------------
library(runjags)
library(coda)
library(mcmcplots)
library(stringr)
library(plyr)
library(dplyr)
library(xtable)
library(ggplot2)

source("functions/plotPost.R")

set.seed(12345)

# Convenience function to generate the type of column names Jags outputs.
col_name <- function(name, ...) {
  paste0(name, "[", paste(..., sep=",") , "]")
}

## ----raw_data, cache = 2, dependson = "scan_data_dir"-------------------------
from_year <- 2014
to_year <- 2019
source(paste0("functions/Import_Data_Eredivisie.R"))

## ----processed_data, cache = 2, dependson = "raw_data"------------------------
eredivisie <- eredivisie %>%
  mutate(MatchResult = sign(HomeGoals - AwayGoals)) # -1 Away win, 0 Draw, 1 Home win

# Creating a data frame d with only the complete match results
d <- na.omit(eredivisie)

# Lists with the unique names of all teams and all seasons in the database
teams <- unique(c(d$HomeTeam, d$AwayTeam))
seasons <- unique(d$Season)

# A list for JAGS with the data from d where the strings are coded as integers
data_list <- list(HomeGoals = d$HomeGoals, AwayGoals = d$AwayGoals, 
                  HomeTeam = as.numeric(factor(d$HomeTeam, levels=teams)),
                  AwayTeam = as.numeric(factor(d$AwayTeam, levels=teams)),
                  Season = as.numeric(factor(d$Season, levels=seasons)),
                  n_teams = length(teams), n_games = nrow(d), 
                  n_seasons = length(seasons))

## ----participation_by_season, fig.height=5, fig.width=7-----------------------
qplot(Season, HomeTeam, data=d, ylab="Team", xlab = "Season")

## ----jags_model_description, cache=2, tidy=FALSE------------------------------
m3_string <- "model {
for(i in 1:n_games) {
HomeGoals[i] ~ dpois(lambda_home[Season[i], HomeTeam[i],AwayTeam[i]])
AwayGoals[i] ~ dpois(lambda_away[Season[i], HomeTeam[i],AwayTeam[i]])
}

for(season_i in 1:n_seasons) {
for(home_i in 1:n_teams) {
for(away_i in 1:n_teams) {
lambda_home[season_i, home_i, away_i] <- exp( home_baseline[season_i] + skill[season_i, home_i] - skill[season_i, away_i])
lambda_away[season_i, home_i, away_i] <- exp( away_baseline[season_i] + skill[season_i, away_i] - skill[season_i, home_i])
}
}
}

skill[1, 1] <- 0 
for(j in 2:n_teams) {
skill[1, j] ~ dnorm(group_skill, group_tau)
}

group_skill ~ dnorm(0, 0.0625)
group_tau <- 1/pow(group_sigma, 2)
group_sigma ~ dunif(0, 3)

home_baseline[1] ~ dnorm(0, 0.0625)
away_baseline[1] ~ dnorm(0, 0.0625)

for(season_i in 2:n_seasons) {
skill[season_i, 1] <- 0 
for(j in 2:n_teams) {
skill[season_i, j] ~ dnorm(skill[season_i - 1, j], season_tau)
}
home_baseline[season_i] ~ dnorm(home_baseline[season_i - 1], season_tau)
away_baseline[season_i] ~ dnorm(away_baseline[season_i - 1], season_tau)
}

season_tau <- 1/pow(season_sigma, 2) 
season_sigma ~ dunif(0, 3) 
}"

## ----monte_carlo, cache=2, dependson=list("processed_data","jags_model_description"), results='hide'----
# Compiling the model
m3 <- run.jags(method="parallel",
               model=m3_string,
               monitor=c("home_baseline", "away_baseline","skill", "season_sigma", "group_sigma", "group_skill"),
               data=data_list,
               n.chains=3,
               adapt=10000,
               burnin=10000,
               sample=15000,
               thin=8,
               summarise=FALSE,
               plots=FALSE)
# Generating MCMC samples
s3 <- as.mcmc.list(m3)
# Merging the three MCMC chains into one matrix
ms3 <- as.matrix(s3)

## ----mu_sigma_params, fig.height=3--------------------------------------------
plot(s3[, "group_skill"])
plot(s3[, "group_sigma"])
plot(s3[, "season_sigma"])

## ----overall_home_advantage---------------------------------------------------
plotPost(exp(ms3[,col_name("home_baseline",to_year-from_year)]) - exp(ms3[,col_name("away_baseline",to_year-from_year)]), compVal = 0, xlab = "Home advantage in number of goals")

## ----model_predictions, cache=2, dependson="monte_carlo", tidy=FALSE----------
n <- nrow(ms3)
m3_pred <- sapply(1:nrow(eredivisie), function(i) {
  home_team <- which(teams == eredivisie$HomeTeam[i])
  away_team <- which(teams == eredivisie$AwayTeam[i])
  season <- which(seasons == eredivisie$Season[i])
  home_skill <- ms3[, col_name("skill", season, home_team)] 
  away_skill <- ms3[, col_name("skill", season, away_team)]
  home_baseline <- ms3[, col_name("home_baseline", season)]
  away_baseline <- ms3[, col_name("away_baseline", season)]
  
  home_goals <- rpois(n, exp(home_baseline + home_skill - away_skill))
  away_goals <- rpois(n, exp(away_baseline + away_skill - home_skill))
  home_goals_table <- table(home_goals)
  away_goals_table <- table(away_goals)
  match_results <- sign(home_goals - away_goals)
  match_results_table <- table(match_results)
  
  mode_home_goal <- as.numeric(names(home_goals_table)[ which.max(home_goals_table)])
  mode_away_goal <- as.numeric(names(away_goals_table)[ which.max(away_goals_table)])
  match_result <-  as.numeric(names(match_results_table)[which.max(match_results_table)])
  rand_i <- sample(seq_along(home_goals), 1)
  
  c(mode_home_goal = mode_home_goal, mode_away_goal = mode_away_goal, match_result = match_result,
    mean_home_goal = mean(home_goals), mean_away_goal = mean(away_goals),
    rand_home_goal = home_goals[rand_i], rand_away_goal = away_goals[rand_i],
    rand_match_result = match_results[rand_i])
})
m3_pred <- t(m3_pred)

## ----hist_home_goal, fig.height=3, fig.width=4--------------------------------
hist(eredivisie$HomeGoals, breaks= (-1:max(eredivisie$HomeGoals, na.rm=TRUE)) + 0.5, xlim=c(-0.5, 10), main = "Distribution of the number of goals\nscored by a home team in a match",
    xlab = "")

## ----mode_home_goal, fig.height=3, fig.width=4--------------------------------
hist(m3_pred[ , "mode_home_goal"], breaks= (-1:max(m3_pred[ , "mode_home_goal"])) + 0.5, xlim=c(-0.5, 10),
    main = "Distribution of predicted most \nprobable score by a home team in\na match",
    xlab = "")

## ----mean_home_goal, fig.height=3, fig.width=4--------------------------------
hist(m3_pred[ , "mean_home_goal"], breaks= (-1:max(m3_pred[ , "mean_home_goal"])) + 0.5, xlim=c(-0.5, 10),
    main = "Distribution of predicted mean \n score by a home team in a match",
    xlab = "")

## ----rand_home_goal, fig.height=3, fig.width=4--------------------------------
hist(m3_pred[ , "rand_home_goal"], breaks= (-1:max(m3_pred[ , "rand_home_goal"])) + 0.5, xlim=c(-0.5, 10),
    main = "Distribution of randomly drawn \n score by a home team in a match",
    xlab = "")

## ----validation_mode_home_goal------------------------------------------------
mean(eredivisie$HomeGoals == m3_pred[ , "mode_home_goal"], na.rm=T)

## ----validation_mean_home_goal------------------------------------------------
mean((eredivisie$HomeGoals - m3_pred[ , "mean_home_goal"])^2, na.rm=T)

## ----hist_actual_match_result, fig.height=3, fig.width=4----------------------
hist(eredivisie$MatchResult, breaks= (-2:1) + 0.5, xlim=c(-1.5, 1.5), ylim=c(0, 1000), main = "Actual match results",
    xlab = "")

## ----hist_pred_match_result, fig.height=3, fig.width=4------------------------
hist(m3_pred[ , "match_result"], breaks= (-2:1) + 0.5, xlim=c(-1.5, 1.5), ylim=c(0, 1000), main = "Predicted match results",
    xlab = "")

## ----hist_rand_match_result, fig.height=3, fig.width=4------------------------
hist(m3_pred[ , "rand_match_result"], breaks= (-2:1) + 0.5, xlim=c(-1.5, 1.5), ylim=c(0, 1000), main = "Randomized match results",
    xlab = "")

## ----validation_match_result--------------------------------------------------
mean(eredivisie$MatchResult == m3_pred[ , "match_result"], na.rm=T)

## ----team_skill, fig.height=8, dpi=90-----------------------------------------
team_skill <- ms3[, str_detect(string=colnames(ms3), paste0("skill\\[",to_year-from_year,","))]
team_skill <- (team_skill - rowMeans(team_skill)) + ms3[, paste0("home_baseline[",to_year-from_year,"]")]
team_skill <- exp(team_skill)
colnames(team_skill) <- teams
team_skill <- team_skill[,order(colMeans(team_skill), decreasing=T)]
old_par <- par(mar=c(2,0.7,0.7,0.7), xaxs='i')
caterplot(team_skill, labels.loc="above", val.lim=c(0.7, 3.8))
par(old_par)

## ----team_skill_PSV_Ajax, fig.height=3, fig.width=7---------------------------
plotPost(team_skill[, "Ajax"] - team_skill[, "PSV Eindhoven"], compVal = 0, xlab = "<- PSV     vs     Ajax ->")

## ----predict_future_matches, results='asis'-----------------------------------
eredivisie_forecast <- eredivisie[is.na(eredivisie$HomeGoals), c("Season", "Week", "HomeTeam", "AwayTeam")]
m3_forecast <- m3_pred[is.na(eredivisie$HomeGoals),] 
eredivisie_forecast$mean_home_goals <- round(m3_forecast[,"mean_home_goal"], 1) 
eredivisie_forecast$mean_away_goals <- round(m3_forecast[,"mean_away_goal"], 1)
eredivisie_forecast$mode_home_goals <- m3_forecast[,"mode_home_goal"] 
eredivisie_forecast$mode_away_goals <- m3_forecast[,"mode_away_goal"]
eredivisie_forecast$predicted_winner <- ifelse(m3_forecast[ , "match_result"] == 1, eredivisie_forecast$HomeTeam, 
                                           ifelse(m3_forecast[ , "match_result"] == -1, eredivisie_forecast$AwayTeam, "Draw"))

rownames(eredivisie_forecast) <- NULL
print(xtable(eredivisie_forecast, align="cccccccccc"), type="html")

## ----predict_past_matches, results='asis'-------------------------------------
eredivisie_sim <- eredivisie[is.na(eredivisie$HomeGoals), c("Season", "Week", "HomeTeam", "AwayTeam")]
eredivisie_sim$home_goals <- m3_forecast[,"rand_home_goal"] 
eredivisie_sim$away_goals <- m3_forecast[,"rand_away_goal"]
eredivisie_sim$winner <- ifelse(m3_forecast[ , "rand_match_result"] == 1, eredivisie_forecast$HomeTeam, 
                            ifelse(m3_forecast[ , "rand_match_result"] == -1, eredivisie_forecast$AwayTeam, "Draw"))

rownames(eredivisie_sim) <- NULL
print(xtable(eredivisie_sim, align="cccccccc"), type="html")

## ----function_plot_goals------------------------------------------------------
plot_goals <- function(home_goals, away_goals) { 
  old_par <- par(mar = c(0, 0, 0, 0))
  par(mfrow = c(2, 2), mar=rep(2.2, 4))
  
	n_matches <- length(home_goals) 
	goal_diff <- home_goals - away_goals 
	match_result <- ifelse(goal_diff < 0, "away_win", ifelse(goal_diff > 0, "home_win", "equal")) 
	hist(home_goals, xlim = c(-0.5, 10), breaks = (0:100) - 0.5)
	hist(away_goals, xlim = c(-0.5, 10), breaks = (0:100) - 0.5) 
	hist(goal_diff, xlim = c(-6, 6), breaks = (-100:100) - 0.5)
	barplot(table(match_result)/n_matches, ylim = c(0, 1))
	par(old_par)
}

## ----setup_top2teams_comparison-----------------------------------------------
n <- nrow(ms3)
home_team <- which(teams == "AZ Alkmaar")
away_team <- which(teams == "PSV Eindhoven")
season <- which(seasons == paste0(to_year-1,"/",to_year))
home_skill <- ms3[, col_name("skill", season, home_team)] 
away_skill <- ms3[, col_name("skill", season, away_team)]
home_baseline <- ms3[, col_name("home_baseline", season)]
away_baseline <- ms3[, col_name("away_baseline", season)]

home_goals <- rpois(n, exp(home_baseline + home_skill - away_skill))
away_goals <- rpois(n, exp(away_baseline + away_skill - home_skill))

## ----plot_goals, fig.height=5-------------------------------------------------
old_par <- par(mfrow = c(2, 2), mar=rep(2.2, 4))
plot_goals(home_goals, away_goals)
par(old_par)

## -----------------------------------------------------------------------------
1 / c(AZ =  mean(home_goals > away_goals), Draw = mean(home_goals == away_goals), PSV = mean(home_goals < away_goals))

## ----payouts, results='asis'--------------------------------------------------
goals_payout <- laply(0:6, function(home_goal) {
  laply(0:6, function(away_goal) {
    1 / mean(home_goals == home_goal & away_goals  == away_goal)
  })
})

colnames(goals_payout) <- paste("PSV Eindhoven", 0:6, sep=" - ")
rownames(goals_payout) <- paste("AZ Alkmaar", 0:6, sep=" - ")
goals_payout <- round(goals_payout, 1)
print(xtable(goals_payout, align="cccccccc"), type="html")

